design/open/repository-operations-error-handling.md

```diff
@@ -0,0 +1,121 @@
+  * Once all correlated entities were successfully deleted, the `Repository` entry itself will can be deleted too
+  * Any failure to delete any entity during the deletion process, still leaves the DB in consistent state as the `Repository` entity, and by that its correlated entities, are unreachable due to the `deleted` state
+
+![Repository State Transitions](diagrams/repository-state-transitions.png)
+[(excalidraw file)](diagrams/repository-state-transitions.excalidraw)
+
+### Repository Unique Identifier
+Each `Repository` will have a `unique_identifier` (To be discussed later) which can differ it from other repositories with the same **Name** (`RepositoryID`). This will provide us with the ability to differ object correlated with different repositories with the same **Name**, from different points in time. Each repository will have a designated KV partition, identified by its **Name** and `unique_identifier` and all correlated objects will be handled in this partition, making it possible to identify objects correlated only to an `active` `Repository`. Moreover, as the `unique_identifier` is an attribute of the `Repository`, in order to create the partition name, the `Repository` entity must be acquired, and so non-exist or unusable `Repository` entries will make unreachable partitions and, in turn, unreachable entities, making it safe to use any reachable entity
+
+### Cleaning Up an `initial` Repository
+A `Repository` entity with `initial` status is either being created (which will soon change the status to `active`), or already failed to create without changing status to `failed` (due to, for example, a crashed process). We will decide a `Repository` is failed if it is in `initial` state, and a sufficient time from its creation has passed. In that case, we will change the `Repository` state to `failed` and make him a candidate for deletion.
+This can be checked every time the `Repository` is accessed (trivial) and by a designated cleanup procedure (part of the above mentioned GC, or a designated scan at startup).
+ 
+##  Flows
+### CreateRepository
+* Create a `Repository` entity with its given `RepositoryID`, randomly generated `UniqueID` and `State` = `initial`, under the common `graveler` partition
+  * If failed return the error (`graveler.ErrNotUnique` will be returned in case a `Repository` entity with the same `RepositoryID` exists) - entity was not created - no harm done.
+* Generate the partition key for the new `Repository`, from its `RepositoryID` and `unique_identifier`
+* Create the initial `Commit` with a `CommitID` (same as it is done today) and store it with the partition key generated above
+  * If failed - try to set the `Repository` entry state to `failed` and return the error from the KV layer
+  * If cannot set the `Repository` state to `failed` it is still consistent, as it is still in `initial` state
```

Comment:
@itaiad200 - which cleanup protocol are tou refering too? The set to `failed`? It is done to allow this repo to be cleaned in the next cleanup routine (whatever it may be), without waiting for the timeout (from TBD #3) to expire.
Other than that, no cleanup actions are done as part of `CreateRepository`

Reply:
Exactly - it's an optimization that we should introduce at a later stage, since we have to handle stuck `initial` state now anyhow.